### PR TITLE
ARROW-17940: [Java][Gandiva] Implement Reserve for JavaBuffer

### PR DESCRIPTION
**Pull request by js8544**
See previous discussions at https://github.com/apache/arrow/pull/14261

### CODE REVIEW

**Comment by js8544** [comment]
This PR is based on changes from https://github.com/apache/arrow/pull/14310 so will need to wait for that to merge.

File: java/gandiva/src/main/cpp/jni_common.cc [hunk 1]
```diff
@@ -15,6 +15,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
+#include <arrow/status.h>
```
Comment (pitrou) [inline java/gandiva/src/main/cpp/jni_common.cc:18]: Can you move this include with the other Arrow includes below?

Comment (js8544) [reply java/gandiva/src/main/cpp/jni_common.cc]: done

File: java/gandiva/src/main/cpp/jni_common.cc [hunk 2]
```diff
@@ -744,14 +733,29 @@ Status JavaResizableBuffer::Resize(const int64_t new_size, bool shrink_to_fit) {
...
   capacity_ = ret_capacity;
   return Status::OK();
 }
 
+Status JavaResizableBuffer::Resize(const int64_t new_size, bool shrink_to_fit) {
+  if (shrink_to_fit == true) {
+    return Status::NotImplemented("shrink not implemented");
+  }
+
+  if (ARROW_PREDICT_TRUE(new_size < capacity())) {
```
[the remark this answers is not in the document]
Comment (js8544) [reply java/gandiva/src/main/cpp/jni_common.cc]: done

File: java/gandiva/src/main/cpp/jni_common.cc [hunk 3]
```diff
@@ -744,14 +733,29 @@ Status JavaResizableBuffer::Resize(const int64_t new_size, bool shrink_to_fit) {
...
-  DCHECK_GE(ret_capacity, new_size);
 
   data_ = reinterpret_cast<uint8_t*>(ret_address);
-  size_ = new_size;
   capacity_ = ret_capacity;
   return Status::OK();
 }
 
+Status JavaResizableBuffer::Resize(const int64_t new_size, bool shrink_to_fit) {
+  if (shrink_to_fit == true) {
```
[the remark this answers is not in the document]
Comment (js8544) [reply java/gandiva/src/main/cpp/jni_common.cc:743]: Yes, it's not needed by gandiva

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Review by pitrou — APPROVED** [review]

**Pull request merged**